File: 3844-number-of-ways-to-assign-edge-weights-i/number-of-ways-to-assign-edge-weights-i.cpp
class Solution {
public:
    int assignEdgeWeights(vector<vector<int>>& edges) {
        int MOD = 1e9 + 7;
        int n = edges.size() + 1;
        vector<vector<int>> memo(n+1 , vector<int> (2 , -1));
        auto dp = [&](this auto&& dp , int i , int rem) -> int {
            if (i == 0) return (int)(rem == 0);
            auto &ret = memo[i][rem];
            if (~ret) return ret;
            ret = (dp(i - 1, rem) + dp(i - 1, rem ^ 1)) % MOD;
            return ret;
        };
        dp(n , 0);
        dp(n , 1);
        // memo has all number of ways to get odd sum of edges wehighting 1 or 2 for mx (number of nodes)
        vector<vector<int>> adj(n+1);
        for (auto v : edges) {
            adj[v[0]].push_back(v[1]);
            adj[v[1]].push_back(v[0]);
        }
        auto dfs = [&](this auto&& dfs , int u , int p) -> int {
            int len = 0;
            for (auto v : adj[u]) {
                if (v == p) continue;
                len = max(len , 1 + dfs(v , u));  
            }
            return len;
        };
        int maxi = dfs(1 , 0);
        // cout << n << ' ' << maxi << ' ' << memo[1][1] << '\n';
        return memo[maxi][1];
    }
};

File: 3844-number-of-ways-to-assign-edge-weights-i/number-of-ways-to-assign-edge-weights-i.java
class Solution {
    static final int MOD = 1_000_000_007;
    private List<List<Integer>> adj;
    private int dfs(int u , int p) {
        int len = 0;
        for (int v : adj.get(u)) {
            if (v == p) continue;
            len = Math.max(len , 1 + dfs(v , u));
        }
        return len;
    }
    static int binPow(int pow) {
        long res = 1 , base = 2;
        while (pow > 0) {
            if ((pow & 1) == 1) res = res * base % MOD;
            base = base * base % MOD;
            pow >>= 1;
        }
        return (int)res;
    }
    public int assignEdgeWeights(int[][] edges) {
        int n = edges.length+1;
        adj = new ArrayList<>(n+1);
        for (int i=0 ; i<=n ; i++) {
            adj.add(new ArrayList<>());
        }
        for (var v : edges) {
            adj.get(v[0]).add(v[1]);
            adj.get(v[1]).add(v[0]);
        }
        int maxi = dfs(1 , 0);
        return binPow(maxi-1);
    }
}